libraries/Microsoft.Bot.Connector/Authentication/AppCredentials.cs

```diff
@@ -0,0 +1,232 @@
+        public static void TrustServiceUrl(string serviceUrl, DateTime expirationTime)
+        {
+            lock (TrustedHostNames)
+            {
+                TrustedHostNames[new Uri(serviceUrl).Host] = expirationTime;
+            }
+        }
+
+        /// <summary>
+        /// Checks if the service url is for a trusted host or not.
+        /// </summary>
+        /// <param name="serviceUrl">The service url.</param>
+        /// <returns>True if the host of the service url is trusted; False otherwise.</returns>
+        public static bool IsTrustedServiceUrl(string serviceUrl)
+        {
+            if (Uri.TryCreate(serviceUrl, UriKind.Absolute, out Uri uri))
+            {
+                return IsTrustedUrl(uri);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Apply the credentials to the HTTP request.
+        /// </summary>
+        /// <param name="request">The HTTP request.</param><param name="cancellationToken">Cancellation token.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public override async Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (ShouldSetToken(request))
+            {
+                string token = await this.GetTokenAsync().ConfigureAwait(false);
+                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
+            }
+
+            await base.ProcessHttpRequestAsync(request, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Gets an OAuth access token.
+        /// </summary>
+        /// <param name="forceRefresh">True to force a refresh of the token; or false to get
+        /// a cached token if it exists.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        /// <remarks>If the task is successful, the result contains the access token string.</remarks>
+        public async Task<string> GetTokenAsync(bool forceRefresh = false)
+        {
+            var token = await authenticator.Value.GetTokenAsync(forceRefresh).ConfigureAwait(false);
+            return token.AccessToken;
+        }
+
+        /// <summary>
+        /// Builds the lazy <see cref="AdalAuthenticator" /> to be used for token acquisition.
+        /// </summary>
+        /// <returns>A lazy <see cref="AdalAuthenticator"/>.</returns>
+        protected abstract Lazy<AdalAuthenticator> BuildAuthenticator();
+
+        private static bool IsTrustedUrl(Uri uri)
+        {
+            lock (TrustedHostNames)
+            {
+                if (TrustedHostNames.TryGetValue(uri.Host, out DateTime trustedServiceUrlExpiration))
+                {
+                    // check if the trusted service url is still valid
+                    if (trustedServiceUrlExpiration > DateTime.UtcNow.Subtract(TimeSpan.FromMinutes(5)))
```

Comment:
I'm not sure why "5 minutes" is a magic number. Wouldn't it simply be >=  for the true/false check?